components/cam/src/physics/cam/micro_p3.F90

```diff
@@ -346,6 +346,121 @@ END SUBROUTINE p3_init_b
+         acn, qv, th, qc, nc, qr, nr, qitot, nitot, qirim, birim, xxlv, xxls, xlf, qc_incld, qr_incld, qitot_incld, &
+         qirim_incld, nc_incld, nr_incld, nitot_incld, birim_incld
+
+    logical(btype), intent(inout) :: log_nucleationPossible, log_hydrometeorsPresent
+
+    ! locals
+    integer :: k
+    real(rtype) :: dum
+
+    k_loop_1: do k = kbot,ktop,kdir
+
+       !calculate some time-varying atmospheric variables
+       !AaronDonahue - changed "rho" to be defined on nonhydrostatic
+       !assumption, consistent with pressure based coordinate system
+       !             - moved latent heat calculation to above.  Latent
+       !heat is determined by calling a p3_util function so that it
+       !can be made consistent with E3SM definition of latent heat
+       rho(k)     = pdel(k)/dzq(k)/g  ! pres(k)/(rd*t(k))
+       inv_rho(k) = 1._rtype/rho(k)
+       qvs(k)     = qv_sat(t(k),pres(k),0)
+       qvi(k)     = qv_sat(t(k),pres(k),1)
+
+       sup(k)     = qv(k)/qvs(k)-1._rtype
+       supi(k)    = qv(k)/qvi(k)-1._rtype
+
+       rhofacr(k) = (rhosur*inv_rho(k))**0.54_rtype
+       rhofaci(k) = (rhosui*inv_rho(k))**0.54_rtype
+       dum          = 1.496e-6_rtype*t(k)**1.5_rtype/(t(k)+120._rtype)  ! this is mu
+       acn(k)     = g*rhow/(18._rtype*dum)  ! 'a' parameter for droplet fallspeed (Stokes' law)
+
+       !specify cloud droplet number (for 1-moment version)
+       if (.not.(log_predictNc)) then
+          nc(k) = nccnst*inv_rho(k)
+       endif
+
+       if ((t(k).lt.zerodegc .and. supi(k).ge.-0.05_rtype) .or.                              &
+            (t(k).ge.zerodegc .and. sup(k).ge.-0.05_rtype )) log_nucleationPossible = .true.
+
+       !--- apply mass clipping if dry and mass is sufficiently small
+       !    (implying all mass is expected to evaporate/sublimate in one time step)
+
+       if (qc(k).lt.qsmall .or. (qc(k).lt.1.e-8_rtype .and. sup(k).lt.-0.1_rtype)) then
+          qv(k) = qv(k) + qc(k)
+          th(k) = th(k) - exner(k)*qc(k)*xxlv(k)*inv_cp
+          qc(k) = 0._rtype
+          nc(k) = 0._rtype
+       else
+          log_hydrometeorsPresent = .true.    ! updated further down
+       endif
+
+       if (qr(k).lt.qsmall .or. (qr(k).lt.1.e-8_rtype .and. sup(k).lt.-0.1_rtype)) then
+          qv(k) = qv(k) + qr(k)
+          th(k) = th(k) - exner(k)*qr(k)*xxlv(k)*inv_cp
+          qr(k) = 0._rtype
+          nr(k) = 0._rtype
+       else
+          log_hydrometeorsPresent = .true.    ! updated further down
+       endif
+
+       if (qitot(k).lt.qsmall .or. (qitot(k).lt.1.e-8_rtype .and.             &
+            supi(k).lt.-0.1_rtype)) then
+          qv(k) = qv(k) + qitot(k)
+          th(k) = th(k) - exner(k)*qitot(k)*xxls(k)*inv_cp
+          qitot(k) = 0._rtype
+          nitot(k) = 0._rtype
+          qirim(k) = 0._rtype
+          birim(k) = 0._rtype
+       else
+          log_hydrometeorsPresent = .true.    ! final update
+       endif
```

Comment:
I wonder if we could turn these into a general purpose subroutine too since they are fairly repetitive.

Reply:
Probably, but I'm not going to worry about it now.

Reply:
sounds good to me.